payment/src/main/java/org/killbill/billing/payment/core/janitor/IncompletePaymentAttemptTask.java

```diff
@@ -155,18 +256,135 @@ public boolean apply(final PaymentTransactionModelDao input) {
+        try {
+            return updatePaymentAndTransactionIfNeeded(accountId,
+                                                       paymentTransactionId,
+                                                       currentTransactionStatus,
+                                                       paymentTransactionInfoPlugin,
+                                                       null,
+                                                       null,
+                                                       internalTenantContext);
+        } catch (final LockFailedException e) {
+            log.warn("Error locking accountRecordId='{}'", internalTenantContext.getAccountRecordId(), e);
+            return false;
+        }
+    }
+
+    private boolean updatePaymentAndTransactionIfNeeded(final UUID accountId,
+                                                        final UUID paymentTransactionId,
+                                                        @Nullable final TransactionStatus currentTransactionStatus,
+                                                        @Nullable final PaymentTransactionInfoPlugin paymentTransactionInfoPlugin,
+                                                        @Nullable final Integer attemptNumber,
+                                                        @Nullable final UUID userToken,
+                                                        final InternalTenantContext internalTenantContext) throws LockFailedException {
+        // First, fix the transaction itself
+        final TransactionStatus latestTransactionStatus = incompletePaymentTransactionTask.updatePaymentAndTransactionIfNeeded(accountId,
+                                                                                                                               paymentTransactionId,
+                                                                                                                               currentTransactionStatus,
+                                                                                                                               paymentTransactionInfoPlugin,
+                                                                                                                               attemptNumber,
+                                                                                                                               userToken,
+                                                                                                                               internalTenantContext);
+        final boolean hasTransactionChanged = latestTransactionStatus == null;
+
+        // Don't insert a notification for the on-the-fly Janitor
+        final boolean shouldInsertNotification = attemptNumber != null;
+        if (!hasTransactionChanged && shouldInsertNotification) {
+            insertNewNotificationForUnresolvedTransactionIfNeeded(paymentTransactionId,
```

Comment:
Is this really necessary since we have the main `doIteration` loop ?

Reply:
Yes: `doIteration` is only called by the in-memory task (for the control payments stuck in an `INIT` state only).